File: src/main/java/org/example/linkedlist/ReorderLL.java
package org.example.linkedlist;

public class ReorderLL {

    public void reorderList(ListNode head) {
        ListNode slow = head, fast = head.next;
        if (slow.next == null) return;

        // To find the middle point
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // Reverse the second half of the LL
        ListNode second = slow.next;
        slow.next = null;
        ListNode prev = null, temp;
        while (second != null) {
            temp = second.next;
            second.next = prev;
            prev = second;
            second = temp;
        }

        second = prev;
        ListNode first = head;
        while (second != null) {
            ListNode temp1 = first.next, temp2 = second.next;
            first.next = second;
            second.next = temp1;
            first = temp1;
            second = temp2;
        }
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        ListNode sec = new ListNode(2);
        ListNode third = new ListNode(3);
        ListNode fourth = new ListNode(4);
        ListNode fifth = new ListNode(5);
        head.next = sec;
        sec.next = third;
        third.next = fourth;
        fourth.next = fifth;

        new ReorderLL().reorderList(head);
        System.out.println(head);
    }
}

File: src/main/java/org/example/linkedlist/ReorderLL.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from typing import Optional

from org.example.linkedlist.ListNode import ListNode


class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # find middle
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # reverse second half
        second = slow.next
        # sever the link from mid to create two separate list
        prev = slow.next = None
        while second:
            tmp = second.next
            second.next = prev
            prev = second
            second = tmp

        # merge two halves
        # prev points to head of the reversed list
        first, second = head, prev
        while second:  # we choose second list as len of second is <= first
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first, second = tmp1, tmp2